lib/opencensus/trace/span.rb

```diff
@@ -0,0 +1,198 @@
+# Copyright 2017 OpenCensus Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+module OpenCensus
+  module Trace
+    ##
+    # Span represents a span in a trace record. Spans are contained in
+    # a trace and arranged in a forest. That is, each span may be a root span
+    # or have a parent span, and may have zero or more children.
+    #
+    class Span
+      ##
+      # A unique identifier for a trace. All spans from the same trace share
+      # the same `trace_id`. The ID is a 16-byte value represented as a
+      # hexadecimal string.
+      #
+      # @return [String]
+      #
+      attr_reader :trace_id
+
+      ##
+      # A unique identifier for a span within a trace, assigned when the span
+      # is created. The ID is an 8-byte value represented as a hexadecimal
+      # string.
+      #
+      # @return [String]
+      #
+      attr_reader :span_id
+
+      ##
+      # The `span_id` of this span's parent span. If this is a root span, then
+      # this field must be empty. The ID is an 8-byte value represented as a
+      # hexadecimal string.
+      #
+      # @return [String]
+      #
+      attr_reader :parent_span_id
+
+      ##
+      # The name of this span.
+      #
+      # @return [OpenCensus::Trace::TruncatableString]
+      #
+      attr_reader :name
+
+      ##
+      # The starting timestamp of this span in UTC, or `nil` if the
```

Comment:
I think by the time we get here (read-only objects for the exporters) the span should always have both a starting and ending timestamp, so those should never be nil.

Reply:
I'm thinking that a SpanBuilder would have a to_span method that would return the read only equivalent of itself. If we treat it as a private method, then it only would be called when the spans are ready for export and should be finished. If it's callable by the user, then it's possible that the end_time may still be nil.

In either case, start_time here should always be set.

Reply:
I was thinking such a method should be private, to communicate the intent SpanBuilder not being readable—and even if it is public, calling to_span on an unfinished SpanBuilder should just raise an exception.

Reply:
Ok, sounds reasonable.

Reply:
In that case, I should make more of the span's fields required if we're assuming the span should be ready for export by this point.

Reply:
Can you remove the "nil" phrases in the comments for this (and for end_time) since these are required?